proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,626 @@
+        }
+
+  let
+    -- e.g. from user input
+    isAdult = True
+    nameFilter = SqlText "A%"
+
+    -- build where clause
+    whereClauses =
+      concat
+        [ ["age > 18" | isAdult]
+        , [s"name ILIKE ${nameFilter}"]
+        ]
+    conjoin cs = mconcat $ intersperse " AND " (cs :: [SqlQuery])
+
+  s"SELECT * FROM tab WHERE ${conjoin whereClauses}"
+    == SqlQuery
+        { sqlText = "SELECT * FROM tab WHERE age > 18 AND name ILIKE ?"
+        , sqlValues = [SqlText "A%"]
+        }
+
+Custom interpolator: HTML
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Imagine a library implements a new ``Html`` type like:
+
+::
+
+  newtype Html = Html Text
+    deriving newtype (Show, IsString, Semigroup, Monoid)
+
+  escapeHtml :: Text -> Text
+  escapeHtml = Text.replace "<" "&lt;" . Text.replace ">" "&gt;"
+
+  newtype RawHtml = RawHtml {unRawHtml :: Text}
+
+That library could define the following instances:
+
+::
+
+  newtype HtmlBuilder = HtmlBuilder (Endo Html)
+    deriving newtype (Semigroup, Monoid)
+
+  instance Buildable Html where
+    type Builder Html = HtmlBuilder
+    toBuilder s = HtmlBuilder (Endo (s <>))
+    fromBuilder (HtmlBuilder (Endo f)) = f mempty
+
+  instance Interpolate String Html where
+    interpolate = interpolate . Text.pack
+  instance Interpolate Text Html where
+    interpolate = toBuilder . Html . escapeHtml
+  instance Interpolate RawHtml Html where
+    interpolate = toBuilder . Html . unRawHtml
+  instance {-# OVERLAPPABLE #-} Show a => Interpolate a Html where
+    interpolate = interpolate . show
+
+And gain access to safe string interpolation with HTML escaping by default:
+
+::
+
+  let title = "Why is 1 > 0?" :: Text
+  let body = "<p>Hello world</p>" :: Text
+
+  s"<h1>${title}</h1>${RawHtml body}"
+    == Html "<h1>Why is 1 &gt; 0?</h1><p>Hello world</p>"
+
+Custom interpolatable type: BigDecimal
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Imagine a library implements a new ``BigDecimal`` type:
+
+::
+
+  data BigDecimal = BigDecimal Integer Int
+
+  renderBigDecimal :: BigDecimal -> String
+  renderBigDecimal (BigDecimal digits scale) =
+    let (int, frac) = splitAt scale (show digits)
+     in int <> "." <> frac
+
+That library could define the following instances:
+
+::
+
+  instance Interpolate BigDecimal String where
+    interpolate = interpolate . renderBigDecimal
+  instance Interpolate BigDecimal Text where
+    interpolate = interpolate . Text.pack . renderBigDecimal
+
+And be able to use it in interpolated strings:
+
+::
+
+  let n = BigDecimal 123456 3
+  s"123456 / 10^3 = ${n}" == "123456 / 10^3 = 123.456"
+
+Format specifiers
+~~~~~~~~~~~~~~~~~
+
+Python is famous for being able to specify format specifiers when interpolating values:
+
+.. code-block:: python
+
+  x = 1.2
+  f"{x:.3f}" == "1.200"
+
+This can be done in a couple ways in user-land:
+
+* Provide functions that format ``Int`` to ``String``, then interpolate that string
+* Add custom data types that implement ``Interpolate``
```

Comment:
To be clear, this still doesn't satisfy what I'm looking for from this section. Here's a stab at what I'd like to see. (Continuing on from after the code block.)

---

This would be provided by libraries, and the design and implementation of those libraries is not specified here. But from a user perspective, here's one possible way such a library could be used:

```
let today :: Day = fromGregorian 2024 08 12
 in s"Today's date is ${Fmt "%a, %d %b %Y" today}."

let earned = -13.2 :: Float
    total = 127.978 :: Float
 in s"""
      Points earned: ${Fmt "+8.2" earned}
      Current total: ${Fmt "+8.2" total}
    """
```

Where these would return the strings:

```
Today's date is Mon, 12 Aug 2024.

Points earned:   -13.20
Current total:  +127.98
```

Reply:
Note that with my proposal you just need to replace your `Fmt` with a function `fmt`! All very straightforward!